# Data Wrangling on Auralin Dataset

Auralin dataset is a fabricated clinical trial dataset for the purpose of demonstrating how to undertake a full data wrangling process systematically.

More information about the data...

- This dataset was constructed with the consultation of real doctors to ensure plausibility.
- This clinical trial data for an alternative insulin was inspired and closely mimics this real [clinical trial for a new inhaled insulin called Afrezza](http://care.diabetesjournals.org/content/38/12/2266.long).
- The data quality issues in this dataset mimic real, [common data quality issues in healthcare data](http://media.hypersites.com/clients/1446/filemanager/Articles/DocCenter_Problem_with_data.pdf). These issues impact quality of care, patient registration, and revenue.
- The patients in this dataset were created using this [fake name generator](http://www.fakenamegenerator.com/order.php) and do not include real names, addresses, phone numbers, emails, etc.

---

## Assess

- Visualize by scrolling
- Visualize by programming

In [1]:
pwd

'/Users/alejandrosanz/Downloads'

In [2]:
import os 
os.chdir('projects_on_GitHub/data_wrangling/diabetes_data_wrangling')

In [4]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

In [10]:
patients.head(50)

patient_id assigned_sex  given_name       surname  \
0            1       female         Zoe       Wellish   
1            2       female      Pamela          Hill   
2            3         male         Jae        Debord   
3            4         male        Liêm          Phan   
4            5         male         Tim       Neudorf   
5            6         male      Rafael         Costa   
6            7       female        Mary         Adams   
7            8       female      Xiuxiu         Chang   
8            9         male       Dsvid    Gustafsson   
9           10       female      Sophie       Cabrera   
10          11       female       Sandy    Gunnarsson   
11          12         male   Abdul-Nur           Isa   
12          13         male  Omeokachie      Ibeamaka   
13          14       female    Anenechi         Chidi   
14          15       female        Asia       Woźniak   
15          16         male       Søren          Lund   
16          17       female         Tám          Liễu   
17          18       female     Roxanne     Andreyeva   
18          19         male     William         Oates   
19          20         male         Zak         Kelly   
20          21       female       Sofia       Karlsen   
21          22         male      Samúel  Guðbrandsson   
22          23         male      Manchu            Su   
23          24         male       Lovre         Galić   
24          25         male       Jakob      Jakobsen   
25          26         male      Gregor          Bole   
26          27       female        Ella          Lund   
27          28         male      Joseph        Tucker   
28          29         male      Robert          Wolf   
29          30         male        Jake      Jakobsen   
30          31       female      Janina      Bazarova   
31          32         male       Tomáš      Navrátil   
32          33       female         Một          Phạm   
33          34         male       Habte         Kifle   
34          35       female     Mariana         Souza   
35          36       female      Kamila      Pecinová   
36          37       female        Naja       Enoksen   
37          38       female     Albinca       Komavec   
38          39       female       Genet      Fesahaye   
39          40       female    Ganimete       Ščančar   
40          41         male    Tješimir         Lukić   
41          42         male      Fakhri      Fakhoury   
42          43       female      Mizuki         Iwata   
43          44       female        Zizi       Schmitt   
44          45       female      Blanka      Jurković   
45          46       female       Liisa       Seppälä   
46          47         male      Javier        Moquin   
47          48       female         Ana       Correia   
48          49       female      Annika         Vaara   
49          50       female      Vivien        Zsinkó   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
5           1140 Willis Avenue      Daytona Beach     Florida   32114.0   
6             3145 Sheila Lane            Burbank          NV   84728.0   
7   2687 Black Oak Hollow Road        Morgan Hill          CA   95037.0   
8           1790 Nutter Street        Kansas City          MO   64105.0   
9            3303 Anmoore Road           New York    New York   10011.0   
10          87 Wood Duck Drive            Rudyard          MI   49780.0   
11      1092 Farm Meadow Drive          Brentwood          TN   37027.0   
12          2544 Worley Avenue          Lynchburg          VA   24504.0   
13            826 Broad Street         Birming

In [14]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [66]:
treatments.melt( value_vars=['auralin', 'novodra'], var_name='type', value_name='vals')

type       vals
0    auralin  41u - 48u
1    auralin          -
2    auralin          -
3    auralin  33u - 36u
4    auralin          -
5    auralin          -
6    auralin  37u - 42u
7    auralin  31u - 38u
8    auralin          -
9    auralin  30u - 36u
10   auralin  29u - 36u
11   auralin          -
12   auralin  29u - 38u
13   auralin          -
14   auralin  27u - 37u
15   auralin  55u - 68u
16   auralin  28u - 37u
17   auralin          -
18   auralin          -
19   auralin          -
20   auralin          -
21   auralin  29u - 39u
22   auralin  53u - 60u
23   auralin          -
24   auralin  31u - 41u
25   auralin          -
26   auralin          -
27   auralin          -
28   auralin  42u - 51u
29   auralin  42u - 49u
30   auralin  35u - 39u
31   auralin          -
32   auralin          -
33   auralin          -
34   auralin  47u - 58u
35   auralin          -
36   auralin  45u - 48u
37   auralin  24u - 37u
38   auralin  44u - 55u
39   auralin          -
40   auralin          -
41   auralin          -
42   auralin  37u - 43u
43   auralin          -
44   auralin  43u - 47u
45   auralin          -
46   auralin  36u - 42u
47   auralin          -
48   auralin          -
49   auralin          -
50   auralin          -
51   auralin  29u - 37u
52   auralin  57u - 64u
53   auralin          -
54   auralin  54u - 67u
55   auralin  34u - 42u
56   auralin          -
57   auralin          -
58   auralin  29u - 37u
59   auralin  61u - 67u
60   auralin  37u - 46u
61   auralin  30u - 39u
62   auralin          -
63   auralin  39u - 46u
64   auralin  24u - 32u
65   auralin          -
66   auralin  32u - 42u
67   auralin          -
68   auralin          -
69   auralin  25u - 32u
70   auralin          -
71   auralin          -
72   auralin          -
73   auralin  36u - 46u
74   auralin          -
75   auralin          -
76   auralin          -
77   auralin          -
78   auralin  35u - 40u
79   auralin  27u - 36u
80   auralin          -
81   auralin  43u - 49u
82   auralin  33u - 41u
83   auralin          -
84   auralin  30u - 41u
85   auralin  42u - 49u
86   auralin          -
87   auralin  31u - 45u
88   auralin          -
89   auralin  44u - 55u
90   auralin  34u - 42u
91   auralin          -
92   auralin          -
93   auralin          -
94   auralin          -
95   auralin  39u - 45u
96   auralin          -
97   auralin          -
98   auralin  25u - 31u
99   auralin          -
100  auralin          -
101  auralin          -
102  auralin          -
103  auralin          -
104  auralin  53u - 64u
105  auralin          -
106  auralin          -
107  auralin          -
108  auralin          -
109  auralin  41u - 47u
110  auralin          -
111  auralin          -
112  auralin  39u - 52u
113  auralin  39u - 43u
114  auralin          -
115  auralin  32u - 41u
116  auralin  36u - 44u
117  auralin          -
118  auralin  46u - 58u
119  auralin  36u - 47u
120  auralin  41u - 51u
121  auralin  42u - 48u
122  auralin  50u - 60u
123  auralin          -
124  auralin          -
125  auralin  48u - 60u
126  auralin          -
127  auralin          -
128  auralin          -
129  auralin  63u - 68u
130  auralin          -
131  auralin          -
132  auralin  30u - 38u
133  auralin          -
134  auralin          -
135  auralin  31u - 40u
136  auralin  29u - 36u
137  auralin          -
138  auralin          -
139  auralin  29u - 45u
140  auralin          -
141  auralin  54u - 54u
142  auralin          -
143  auralin  55u - 59u
144  auralin  24u - 32u
145  auralin  46u - 53u
146  auralin  55u - 61u
147  auralin          -
148  auralin          -
149  auralin  29u - 37u
150  auralin  55u - 66u
151  auralin  40u - 53u
152  auralin  39u - 44u
153  auralin          -
154  auralin          -
155  auralin          -
156  auralin          -
157  auralin          -
158  auralin          -
159  auralin  48u - 58u
160  auralin  47u - 57u
161  auralin  40u - 42u
162  auralin  32u - 42u
163  auralin  44u - 54u
164  auralin  62u - 75u
165  auralin  37u - 46u


In [7]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [8]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [9]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [26]:
pd.set_option('display.max_rows', None)
patients.surname.value_counts()

Doe                  6
Jakobsen             3
Taylor               3
Aranda               2
Dratchev             2
Liễu                 2
Hueber               2
Schiavone            2
Berg                 2
Kadyrov              2
Souza                2
Woźniak              2
Tạ                   2
Lương                2
Lâm                  2
Correia              2
Cabrera              2
Tucker               2
Gersten              2
Nilsen               2
Bùi                  2
Collins              2
Ogochukwu            2
Lund                 2
Batukayev            2
Silva                2
Johnson              2
Parker               2
Cindrić              2
Grímsdóttir          2
Kowalczyk            2
Ščančar              1
Maríasson            1
Heimisson            1
Petersen             1
Tománková            1
Bogolyubova          1
Nowakowski           1
Pecinová             1
Priest               1
Eldarkhanov          1
Nyborg               1
Knutsen              1
Jephcott   

In [27]:
patients.address.value_counts()

123 Main Street                  6
2778 North Avenue                2
2476 Fulton Street               2
648 Old Dear Lane                2
2121 Liberty Avenue              1
4155 Raccoon Run                 1
4105 Ferguson Street             1
4988 Lynn Street                 1
3488 Clair Street                1
570 Alpha Avenue                 1
1934 August Lane                 1
34 Hamill Avenue                 1
2886 Straford Park               1
1846 Joseph Street               1
2566 Ingram Street               1
3479 Elm Drive                   1
3781 Hamill Avenue               1
707 Goodwin Avenue               1
2687 Black Oak Hollow Road       1
1810 Hardesty Street             1
2531 Eastland Avenue             1
373 Fantages Way                 1
1962 Cabell Avenue               1
1027 Tenmile Road                1
4500 Myra Street                 1
2020 Gore Street                 1
4141 Davis Place                 1
3650 Graystone Lakes             1
4168 Coventry Court 

In [56]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
208    121.2
0      121.7
467    122.1
218    122.2
225    122.3
404    123.0
397    123.4
424    123.6
483    123.9
306    124.1
278    124.3
120    124.5
363    124.7
185    125.1
423    125.2
419    126.1
494    126.3
232    126.9
214    126.9
342    127.2
280    128.5
184    128.9
256    128.9
17     129.1
422    129.4
44     129.8
206    129.8
377    130.0
181    130.2
354    130.2
309    130.7
22     130.7
213    131.1
143    131.1
62     131.1
268    132.0
359    132.2
362    132.2
454    132.7
187    133.1
31     133.3
56     133.8
57     134.0
163    134.2
345    134.2
458    134.2
430    135.7

In [49]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [51]:
patients.address.duplicated().sum()

19

In [57]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [59]:
patients[patients[['given_name', 'surname']].duplicated()]

patient_id assigned_sex given_name surname          address      city  \
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state  zip_code        country                      contact birthdate  \
229    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
237    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
244    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
251    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
277    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   

     weight  height   bmi  
229   180.0      72  24.4  
237   180.0      72  24.4  
244   180.0      72  24.4  
251   180.0      72  24.4  
277   180.0      72  24.4

In [54]:
patients[patients.address.duplicated()][['given_name', 'surname']]

given_name     surname
29        Jake    Jakobsen
219         Mỹ       Quynh
229       John         Doe
230  Elisabeth     Knudsen
234    Martina   Tománková
237       John         Doe
242       John     O'Brian
244       John         Doe
249   Benjamin      Mehler
251       John         Doe
257        Jin        Kung
264   Wafiyyah      Asfour
269     Flavia  Fiorentino
277       John         Doe
278   Generosa       Cabán
282      Sandy      Taylor
286      Lewis        Webb
296        Chỉ         Lâm
502        Pat     Gersten

In [31]:
treatments.auralin.isnull().sum()

0

In [32]:
treatments.novodra.isnull().sum()

0

### Quality Issues
####  `patients` table
- zip code is a float not a string
- zip code has four digits sometimes
- Tim Neudorf height is 27 instead of 72 in
- full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns)
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- multiple phone number formats
- Default John Doe data
- Multiple records for Jakebsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight


#### `treatments` table
- missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- lowercase given names and surnames
- missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (4s mistaken as 9s)
- Nulls represented as dashed (-) in auralin and novodra

#### `adverse_reactions` table
- lowercase given names and surnames



### Tidiness Issues

#### `patients` table
- contact column should be split into phone and email columns separately
 
#### `treatments` table
- three variables in two columns (should be like treatment, start_dose, end_dose)
- `given_name` and `surname` being redundant (Add `id` column to connect with `patients` table)
 
 
#### `adverse_reactions` table
- `given_name` and `surname` being redundant
- some kind of verbose and should be joined with `treatments` table


In [68]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [69]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [70]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [76]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns

0           patient_id
1         assigned_sex
2           given_name
3              surname
4              address
5                 city
6                state
7             zip_code
8              country
9              contact
10           birthdate
11              weight
12              height
13                 bmi
14          given_name
15             surname
16             auralin
17             novodra
18         hba1c_start
19           hba1c_end
20        hba1c_change
21          given_name
22             surname
23    adverse_reaction
dtype: object

In [75]:
all_columns.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14     True
15     True
16    False
17    False
18    False
19    False
20    False
21     True
22     True
23    False
dtype: bool

In [73]:
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object